src/Qir/Execution/Tools/Executable/QirExecutable.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Quantum.Qir.Tools.Driver;
+using Microsoft.Quantum.Qir.Utility;
+using Microsoft.Quantum.QsCompiler.BondSchemas.EntryPoint;
+
+namespace Microsoft.Quantum.Qir.Tools.Executable
+{
+    /// <summary>
+    /// Base for creating and running QIR-based executables.
+    /// </summary>
+    public abstract class QirExecutable : IQirExecutable
+    {
+        private const string DriverFileName = "driver";
+        private const string BytecodeFileName = "qir.bc";
+
+        public virtual string SourceDirectoryPath => "src";
+        public abstract string DriverFileExtension { get; }
+
+        protected FileInfo ExecutableFile { get; }
+
+        private readonly byte[] qirBytecode;
+        private readonly ILogger logger;
+        private readonly IQuantumExecutableRunner runner;
+        private readonly IQirDriverGenerator driverGenerator;
+        private readonly IQirExecutableGenerator executableGenerator;
+
+        public QirExecutable(FileInfo executableFile, byte[] qirBytecode, ILogger logger, IQirDriverGenerator driverGenerator, IQirExecutableGenerator executableGenerator, IQuantumExecutableRunner runner)
```

Comment:
Following the approach I suggested in this [comment](https://github.com/microsoft/qsharp-runtime/pull/642/files#r618043290), a `public` constructor should only receive `executableFile` and `qirBytecode` as arguments, and a constructor like this one can be `internal`.